cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -1173,5 +1175,168 @@ TEST(Substrait, JoinPlanInvalidKeys) {
           &ext_set));
 }
 
+TEST(Substrait, SerializeRelation) {
+#ifdef _WIN32
+  GTEST_SKIP() << "ARROW-16392: Substrait File URI not supported for Windows";
+#else
+  ExtensionSet ext_set;
+  auto dummy_schema = schema({field("f1", int32()), field("f2", int32())});
+  // creating a dummy dataset using a dummy table
+  auto format = std::make_shared<arrow::dataset::ParquetFileFormat>();
+  auto filesystem = std::make_shared<fs::LocalFileSystem>();
+
+  std::vector<fs::FileInfo> files;
+  const std::vector<std::string> f_paths = {"/tmp/data1.parquet", "/tmp/data2.parquet"};
+
+  for (const auto& f_path : f_paths) {
+    ASSERT_OK_AND_ASSIGN(auto f_file, filesystem->GetFileInfo(f_path));
+    files.push_back(std::move(f_file));
+  }
+
+  ASSERT_OK_AND_ASSIGN(auto ds_factory, dataset::FileSystemDatasetFactory::Make(
+                                            std::move(filesystem), std::move(files),
+                                            std::move(format), {}));
+  ASSERT_OK_AND_ASSIGN(auto dataset, ds_factory->Finish(dummy_schema));
+
+  auto options = std::make_shared<dataset::ScanOptions>();
+  options->projection = compute::project({}, {});
+  auto scan_node_options = dataset::ScanNodeOptions{dataset, options};
+
+  auto scan_declaration = compute::Declaration({"scan", scan_node_options});
+
+  ASSERT_OK_AND_ASSIGN(auto serialized_rel,
+                       SerializeRelation(scan_declaration, &ext_set));
+  ASSERT_OK_AND_ASSIGN(auto deserialized_decl,
+                       DeserializeRelation(*serialized_rel, ext_set));
+
+  auto dataset_comparator = [](std::shared_ptr<dataset::Dataset> ds_lhs,
+                               std::shared_ptr<dataset::Dataset> ds_rhs) -> bool {
+    const auto& fds_lhs = checked_cast<const dataset::FileSystemDataset&>(*ds_lhs);
+    const auto& fds_rhs = checked_cast<const dataset::FileSystemDataset&>(*ds_lhs);
+    const auto& files_lhs = fds_lhs.files();
+    const auto& files_rhs = fds_rhs.files();
+
+    bool cmp_fsize = files_lhs.size() == files_rhs.size();
+    uint64_t fidx = 0;
+    for (const auto& l_file : files_lhs) {
+      if (l_file != files_rhs[fidx]) {
+        return false;
+      }
+      fidx++;
+    }
+    bool cmp_file_format = fds_lhs.format()->Equals(*fds_lhs.format());
+    bool cmp_file_system = fds_lhs.filesystem()->Equals(fds_rhs.filesystem());
+    return cmp_fsize && cmp_file_format && cmp_file_system;
+  };
+
+  auto scan_option_comparator = [dataset_comparator](
+                                    const dataset::ScanNodeOptions& lhs,
+                                    const dataset::ScanNodeOptions& rhs) -> bool {
+    bool cmp_rso = lhs.require_sequenced_output == rhs.require_sequenced_output;
+    bool cmp_ds = dataset_comparator(lhs.dataset, rhs.dataset);
+    return cmp_rso && cmp_ds;
+  };
```

Comment:
cc @westonpace @lidavidm 
Would it be better to add a Equals method for `ScanNodeOptions` and may be even for all the other `ExecNodeOptions`?

Reply:
Yes, let's add Equals where necessary. You can also inherit from [EqualityComparable](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/compare.h)

Reply:
Great, should it be a separate PR or we just include here?

Reply:
Sure I will add it here.

Reply:
@lidavidm 
Regarding the `Equals` methods required, 

For `FileSystemDataset` we need to consider the, `FileFormat` (`Equals` available), `FileSystem` (`Equals` available), `FileFragment` (`Equals` TODO add method straightforward), `Partitioning` (`Equals` TODO needs to Add method for Base class and for child classes) and `FragmentSubtree` (`Equals` TODO add method - straightforward)

Is it still okay to add these change in the same PR?

Reply:
Please note that the current `Equals` lambdas doesn't contain the complete functionality.

Reply:
I don't see why not

Reply:
Or is it satisfactory to just write them as test utils. I assume if these `Equals` are not going to be widely used, we can add it to test utils.

Reply:
Again, it's up to your best judgement. I would have thought these are mostly straightforward implementations, but if not and  you feel like it would complicate things too much, then just leave them as is.

Reply:
I'm not currently deep enough in that code to really give you a full answer off the top of my head

Reply:
Let's add them in the same PR then. Otherwise it would complicate improving this PR simultaneously. It is not very complicated, but for the organisation,  these could have been written as supportive functions separately.

Reply:
On second thoughts, may be for this PR the comparison required is satisfactory with those lambdas. We compare the tabel content in one test using Tables and the other test case just need to compare the configs. 

May be we can create JIRAs to include `Equals` for these classes. 

cc @westonpace

Reply:
It makes sense for these things to have an `Equals` method.  It could even be useful to the user.  However, I don't see any reason we need to add it now.  What you have here is fine also.  If we end up doing the same comparison elsewhere we can also turn the lambdas into methods like `AssertDatasetEquals` inside of a `test_util` file.  So let's just keep this as simple lambdas for the moment.